### PR TITLE
Discord prototype

### PR DESCRIPTION
**Pull request by boris-schickedanz**
The loader is working.... 
it needs a bot-token and a discord channel to work
the bot needs to be invited to the server before starting as well.

the TestBot.js can be used to check if permissions and all were ok -> we need some documentation on how to set it up

not sure if we want this branch on development though... ?

### CODE REVIEW

File: libs/domain/feature-loaders/src/discord/DiscordLoader.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            }
+        );
+    }
+
+    protected mapDiscordResult(cursor?: string) {
+        const channel = this.client.channels.cache.get(
+            this.channelId
+        ) as TextChannel;
+
+        // TODO do we need to await this here at all?
```
Comment (adam-arold) [inline libs/domain/feature-loaders/src/discord/DiscordLoader.ts:106]: We don't. In `loadRaw` we just load the "raw" result (whatever the API is returning). This is how `load` looks like:

```ts
public load(context: LoadContext) {
    return pipe(
        TE.Do,
        TE.bind("loadContext", () => this.preLoad(context)),
        TE.bind("rawResult", ({ loadContext }) =>
            this.loadRaw(loadContext)
        ),
        TE.bindW("mappedResult", ({ rawResult }) => {
            return TE.tryCatch(
                async () => this.mapResult(rawResult),
                (e) => new UnknownError(e)
            );
        }),
        TE.map((params) => {
            return {
                cursor: this.getNextCursor(params),
                data: params.mappedResult,
            };
        })
    );
}
```

The relevant part starts with the `rawResult` binding: this just executes `loadRaw` and returns whatever it returns, and then constructs the `mappedResult` and it also comes with built-in error-handling. You can use that or add your custom one, it doesn't really matter.

So in your case `mapDiscordResult` should probably go into `loadRaw` and you can just call `channel.messages.fetch` there and return whatever is in the response.

**Review by adam-arold — COMMENTED** [review]
Looks good! I've left some comments in the code. I think the best course of action might be to create a new implementation branch from this one and don't merge the prototype branch.

File: libs/domain/feature-loaders/src/discord/DiscordLoader.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    DiscordMessages,
+    DiscordMessage
+> {
+    public info = INFO;
+
+    // Discord doesn't allow 1000 here
+    protected batchSize = 100;
+    protected type = DiscordMessage;
+    protected cadenceConfig = {
+        [ScheduleMode.BACKFILL]: { seconds: 5 },
```
[the remark this answers is not in the document]
Comment (boris-schickedanz) [reply libs/domain/feature-loaders/src/discord/DiscordLoader.ts:62]: It depends on how many people will be using the bot... if it is used for the announcement channel only, for sure not. 

https://discord.com/developers/docs/topics/rate-limits

50 requests per second and <10.000 per 10 minutes from one bot are safe... if we make other DAOs / implementers of the CG use their own bot-token, no problem at all:

https://discord.com/developers/docs/topics/rate-limits

File: libs/domain/feature-loaders/src/discord/DiscordLoader.ts (hunk 1 above)
Comment (boris-schickedanz) [reply libs/domain/feature-loaders/src/discord/DiscordLoader.ts]: refactored codec out to CodecDataLoaderBase

**Review by adam-arold — APPROVED** [review]
LGTM!

**Pull request merged**